### ISSUE TITLE
pkg-config is broken in OpenCV-2.4.0

### ISSUE THREAD

**Issue by mshabunin** [mshabunin/test2#722]
Converted from http://code.opencv.org/issues/1925
```
|| by Hiromasa YOSHIMOTO on 2012-05-12 13:24:00
|| Status: Incomplete
|| Priority: Normal
|| Affected: 2.4.0 - 2.4.8
|| Target: -
|| Category: build/install

pkg-config is broken in OpenCV-2.4.0
===========
OpenCV-2.4.0 has a regression bug in pkg-config support.

The following pkg-config commands return wrong information now.  
- "pkg-config opencv --libs"   
 It returns just filename list of libraries. It must use "-L" and "-l"
linker flags.  
- "pkg-config opencv --libs-only-L"   
- "pkg-config opencv --libs-only-l"   
 no output returned.

I wrote a patch to fix this issue. Please find the attachment.


History
=======
|| by Andrey Kamaev on 2012-05-12 16:18:00
Thanks for the patch but it introduces another regression - versioned
opencv-2.4.0.pc becomes completely broken.

Could you point to any pkg-config documentation saying that it must use
"-L" and "-l" linker flags?
-   Target version deleted (2.4.1)
-   Assignee changed from Alexander Shishkov to Andrey Kamaev


|| by Hordur Johannsson on 2012-05-12 19:30:00
I looked at the pkg-config website
(http://people.freedesktop.org/~dbn/pkg-config-guide.html) and they say
that Libs: should contain the linker flags.  
In the case of gcc the full shared library paths work, e.g

    gcc myapp.cpp -o myapp `pkg-config opencv --libs`

works fine when compiling with gcc, now I'm not sure if that would break
with other compilers.

I most often use cmake for my build system. Then I can use the
FindPkgConfig to find OpenCV,  
it will set OpenCV_LDFLAGS to the shared libraries. Here is in example
how you might use that.

    find_package(PkgConfig REQUIRED)pkg_check_modules(OpenCV opencv)include_directories(${OpenCV_INCLUDE_DIRS})  add_executable(myapp myapp.cpp)target_link_libraries(myapp ${OpenCV_LDFLAGS})


|| by Hiromasa YOSHIMOTO on 2012-05-13 11:10:00
Thanks for your comments. I'll fix my patch later.

I mean that the "linker flags" should use -L and -l flags.

I could not find any explicit specification, but  
pkg-config seems to expect -L and -l in linker flags.  
Because it provides the following two options(*1,*2);

-   --libs-only-L for printing -L/-R part of "--libs"
-   --libs-only-l for printing -l part of "--libs"

Anyway, the issue is opencv.pc has been changed in OpenCV-2.4  
and it breaks backword compatibility. Is there any reason to break it?

*1) see man page of pkg-config or http://linux.die.net/man/1/pkg-config
  
*2) Please search pkg-config and "--libs-only-L" on Internet.  
You will find how these options are used in makefile/autotools/cmake
files.


|| by Hiromasa YOSHIMOTO on 2012-05-14 14:45:00
When I checked versioned opencv-2.4.0.pc,   
I found another issue in shared library versioning.  
See #1928 for details.

I'd like to update my patch when #1928 is closed  
because this pkg-config issue depends on #1928.


|| by Peter Housel on 2012-08-28 21:57:00
An important reason to use -L is that in cross-compilation environments,
pkg-config will prefix -L and -I paths with the path found in
PKG_CONFIG_SYSROOT_DIR. This variable normally points to a target
sysroot directory containing include files, libraries, and pkgconfig
files for the cross-compilation target.


|| by Andrey Kamaev on 2012-10-22 12:48:00
-   Assignee deleted (Andrey Kamaev)


|| by Ray Rashif on 2012-11-05 21:17:00
This is still broken in 2.4.3, and it breaks standard pkg-config usage.
Another patch attached which makes use of CMake's functionality instead
of string replacement.
-   File pkgconfig.patch added


|| by Sebastian Krämer on 2012-11-14 21:03:00
I'm having the same issues with pkg-config on 64 bit linux. I tried your
patch but it doesn't solve the issues for me (although the situation has
improved a bit):  

    $ PKG_CONFIG_PATH="`pwd`" pkg-config --libs opencv
    liblibpng /usr/lib64/libQtTest.so /usr/lib64/libQtCore.so /usr/lib64/libQtGui.so /usr/lib64/libQtOpenGL.so /lib64/libXext.so /lib64/libX11.so /lib64/libICE.so /lib64/libSM.so /lib64/libGL.so /lib64/libGLU.so rt pthread m dl stdc++ /lib64/libz.so -L/home/sebi/src/opencv/lib -lopencv_flann -lopencv_highgui -lopencv_imgproc -lopencv_ml -lopencv_core

  
At least the opencv -l parameters are there now.  
Notice the liblibpng at the beginning.


|| by Vadim Pisarevsky on 2014-11-15 13:21:00
There have been some PR's fixing problems with pkg-config. Hopefully,
the problem has been solved.
-   Status changed from Open to Done
-   Affected version set to 2.4.0 - 2.4.8
-   Operating System set to Linux
-   HW Platform set to Any


|| by Timocheck Jackson on 2015-02-17 23:53:00
The problem is still seems to be present with version 2.4.10:

1.  pkg-config opencv --libs-only-l  
    -lrt -lpthread -lm -ldl
2.  pkg-config opencv --libs-only-L

#

Still no "-L" flags. Not possible to link opencv with other programs.
-   Assignee set to Hiromasa YOSHIMOTO
-   Status changed from Done to Incomplete


|| by Steven Puttemans on 2015-02-19 13:11:00
@Timo

Is it possible that you are using the wrong pkg-config command? I have
been using the command for ages on my Ubuntu system with 2.4.10 and it
works perfectly.  
My command:  

    `pkg-config opencv --libs`
```
